src/sp/it/pl/util/hotkey/Hotkeys.kt

```diff
@@ -0,0 +1,174 @@
+                    }
+                }
+
+                override fun nativeKeyReleased(e: NativeKeyEvent) {
+                    keyCombos.values.forEach { if (it.isPressed) it.release() }
+                }
+
+                override fun nativeKeyTyped(e: NativeKeyEvent) {}
+            }
+            try {
+                GlobalScreen.setEventDispatcher(eventDispatcher)
+                GlobalScreen.addNativeKeyListener(keyListener)
+                GlobalScreen.registerNativeHook()
+                this.keyListener = keyListener
+            } catch (e: NativeHookException) {
+                logger.error("Failed to register global hotkeys", e)
+            }
+
+        }
+    }
+
+    fun stop() {
+        if (isRunning()) {
+            try {
+                logger.info { "Stopping global hotkeys" }
+                GlobalScreen.removeNativeKeyListener(keyListener)
+                keyListener = null
+                GlobalScreen.unregisterNativeHook()
+            } catch (e: NativeHookException) {
+                logger.error(e) { "Failed to unregister global hotkeys" }
+            }
+
+        }
+    }
+
+    fun register(action: Action, keys: String) {
+        val keyString = keys.substringAfterLast('+').trim().toLowerCase().replace('_', ' ')
+        val key = KeyCode.values().find { k -> k.getName().equals(keyString, true) }
+                ?: fail { "No KeyCode for ${action.keys}" }
+        register(
+                action,
+                key,
+                *sequenceOf(KeyCode.ALT, KeyCode.ALT_GRAPH, KeyCode.SHIFT, KeyCode.CONTROL, KeyCode.WINDOWS, KeyCode.COMMAND, KeyCode.META)
+                        .filter { k -> keys.contains(k.name, true) || (k==KeyCode.CONTROL && keys.contains("CTRL", true)) }
+                        .toList().toTypedArray()
+        )
+    }
+
+    fun register(action: Action, key: KeyCode, vararg modifiers: KeyCode) {
+        keyCombos[action.id] = KeyCombo(key, *modifiers)
+    }
+
+    fun unregister(action: Action) {
+        keyCombos.remove(action.id)
+    }
+
+    private inner class KeyCombo {
+        val key: KeyCode
+        val modifier: Int
+        var isPressed = false
+            private set(value) {
+                field = value
+            }
+
+        constructor(key: KeyCode, vararg modifiers: KeyCode) {
+            this.key = key
+            this.modifier = run {
+                var m = 0
+                if (KeyCode.SHIFT in modifiers) m += NativeInputEvent.SHIFT_L_MASK
+                if (KeyCode.CONTROL in modifiers) m += NativeInputEvent.CTRL_L_MASK
+                if (KeyCode.WINDOWS in modifiers || KeyCode.META in modifiers) m += NativeInputEvent.META_L_MASK
+                if (KeyCode.ALT in modifiers) m += NativeInputEvent.ALT_L_MASK
+                if (KeyCode.COMMAND in modifiers) m += NativeInputEvent.META_R_MASK
+                if (KeyCode.ALT_GRAPH in modifiers) m += NativeInputEvent.ALT_R_MASK
```

Comment:
this code looks smelly, but I guess a when clause won't do it either

Reply:
It maybe looks that way to you because it mimics an old style of coding with mutating bit mask variables. Not much we can do here Im afraid. I think this is the most readable there is.

Reply:
Actually, I have a pretty idea for this, should I just commit it to the branch?

Reply:
Try if you think its better than this.

Reply:
Nice implementation. Local functions tend to incur unfavourable readability overheads, but I do find this solution superior.